lib/middleware/groups-required.js

```diff
@@ -4,63 +4,89 @@ var helpers = require('../helpers');
       return res.redirect(302, url);
     }
 
-    // If this user must be a member of all groups, we'll ensure that is the
-    // case.
-    var done = groups.length;
-    var safe = false;
+    function isUserInGroups(callback) {
+      var isInGroup = false;
+      var done = assertGroups.length;
 
-    req.user.getGroups(function (err, grps) {
-      if (err) {
-        logger.info('Could not fetch user ' + req.user.email + '\'s groups.');
-        return helpers.render(req, res, view);
-      }
+      req.user.getGroups(function (err, userGroups) {
+        if (err) {
+          return callback(null, false);
+        }
 
-      // Iterate through each group on the user's account, checking to see
-      // whether or not it's one of the required groups.
-      grps.each(function (group, c) {
-        if (groups.indexOf(group.name) > -1) {
-          if (!all || --done === 0) {
-            safe = true;
-            next();
+        // Iterate through each group on the user's account, checking to see
+        // whether or not it's one of the required groups.
+        userGroups.each(
+          function (group, iterateNext) {
+            if (assertGroups.indexOf(group.name) > -1) {
+              if (!all || --done === 0) {
+                isInGroup = true;
+              }
+            }
+            iterateNext();
+          },
+          function () {
+            callback(null, isInGroup);
           }
+        );
+      });
+    }
+
+    helpers.handleAcceptRequest(req, res, {
+      'application/json': function () {
+        if (!req.user) {
+          return helpers.writeJsonError(res, new Error('You are not authenticated. Please log in to access this resource.'), 401);
         }
-        c();
+
+        isUserInGroups(function (err, isInGroup) {
+          if (err) {
+            return helpers.writeJsonError(res, new Error('An internal error occurred. Please contact support.'), 500);
+          }
+
+          if (!isInGroup) {
+            return helpers.writeJsonError(res, new Error('You do not have sufficient permissions to access this resource.'), 403);
+          }
+
+          next();
+        });
       },
-      // If we get here, it means the user didn't meet the requirements,
-      // so we'll send them to the login page with the ?next querystring set.
-      function () {
-        if (!safe) {
-          logger.info('User ' + req.user.email + ' attempted to access a protected endpoint but did not meet the group check requirements.');
-          helpers.render(req, res, view);
+      'text/html': function () {
+        if (!req.user) {
+          return redirectToLoginPage();
         }
-      });
-    });
+
+        isUserInGroups(function (err, isInGroup) {
+          if (err || !isInGroup) {
+            res.status(403);
+            return helpers.render(req, res, 'unauthorized');
+          }
+
+          next();
+        });
+      }
+    }, redirectToLoginPage);
```

Comment:
This behavior isn't right, that I'm redirected to the login page if I haven't specified an Accept header.  I should still be able to attempt authorization, regardless of the Accept header value.  It's only when we need to render an error that we should be inspecting the Accept header, to determine how to render the error (JSON error response, or redirect to login page).  And if no Accept value is specified, default to the first value in the produces array.  

I understand that dealing with the first value in the produces array is a larger task for this library, and that it may be out of scope for this PR.  But I still think this PR should be modified to still attempt authorization, regardless of Accept type.